[BE] Project manager role

nspc97 opened the pull request:
[BE] Project manager role
https://trello.com/c/mY9KaA6l/99-be-project-manager-role

- Each user should get a PM assigned (not mandatory).
- Admins can assign a PM to the user.
- When the PM account is deleted/deactivated, all the users of this PM get unassigned from him.

Commit:
[BE] Project manager role
https://trello.com/c/mY9KaA6l/99-be-project-manager-role

- Each user should get a PM assigned (not mandatory).
- Admins can assign a PM to the user.
- When the PM account is deleted/deactivated, all the users of this PM get unassigned from him.

MihaiDruta commented on src/Controller/UserController.php:
```diff
@@ -610,4 +611,164 @@ public function removeAvatar(
+
+    /**
+     * Set Project Manager
+     * @Rest\Post("/{userId}/set_pm/{pmId}", requirements={"id"="\d+"})
```
Requirements are not correct

MihaiDruta commented on src/Controller/UserController.php:
```diff
@@ -610,4 +611,164 @@ public function removeAvatar(
+     *     description="This user in not a PM!",
+     *     @SWG\Schema(
+     *     @SWG\Property(property="code", type="integer", example=400),
+     *     @SWG\Property(property="message", type="string"),
```
You can add an example message for this response

MihaiDruta commented on src/Entity/User.php:
```diff
@@ -160,8 +160,15 @@ class User implements UserInterface
     private $createdAt;
 
+    /**
+     * Comment id linked to.
```
Wrong description

Commit:
[BE] Project manager role
https://trello.com/c/mY9KaA6l/99-be-project-manager-role

- Each user should get a PM assigned (not mandatory).
- Admins can assign a PM to the user.
- When the PM account is deleted/deactivated, all the users of this PM get unassigned from him.

SpartakusMd commented on src/Repository/UserRepository.php:
```diff
@@ -60,6 +60,7 @@ public function save(User $newUser): void
     public function delete(User $user): void
     {
         $this->commentRepository->anonymizeUserComments($user);
+        $this->clearProjectManager($user);
```
Rename to `unassignProjectManager`

SpartakusMd commented on src/Controller/UserController.php:
```diff
@@ -610,4 +611,164 @@ public function removeAvatar(
+
+    /**
+     * Set Project Manager
+     * @Rest\Post("/{userId}/set_pm/{pmId}", requirements={"userId"="\d+", "pmId"="\d+"})
```
Rename `set_pm` to `assign-project-manager`. The ID of PM should be present in the body not in the URL.

SpartakusMd commented on src/Controller/UserController.php:
```diff
@@ -610,4 +611,164 @@ public function removeAvatar(
+
+    /**
+     * Set Project Manager role
+     * @Rest\Post("/{id}/set_pm", requirements={"id"="\d+"})
```
Role management should be general. Only admins can assign roles. An admin can promote anyone to PM or to admin and also demote to PM, user.

Commit:
[BE] Project manager role
https://trello.com/c/mY9KaA6l/99-be-project-manager-role

- Changed endpoint set_pm to set_role
- The ID of PM is in the body not in the URL

SpartakusMd commented on src/Controller/UserController.php:
```diff
@@ -613,21 +613,28 @@ public function removeAvatar(
-     * Set Project Manager role
-     * @Rest\Post("/{id}/set_pm", requirements={"id"="\d+"})
+     * Set role
+     * @Rest\Post("/{id}/set_role", requirements={"id"="\d+"})
```
In URLs dash "-" should be used instead of underscore "_".

SpartakusMd commented on src/Controller/UserController.php:
```diff
@@ -662,29 +669,54 @@ public function removeAvatar(
-     * @Rest\Post("/{userId}/set_pm/{pmId}", requirements={"userId"="\d+", "pmId"="\d+"})
-     * @ParamConverter("user", options={"mapping": {"userId" : "id"}})
-     * @ParamConverter("projectManager", options={"mapping": {"pmId" : "id"}})
+     * @Rest\Post("/{id}/assign_project_manager", requirements={"id"="\d+"})
```
In URLs dash "-" should be used instead of underscore "_".

SpartakusMd commented on src/DTO/UserDTO.php:
```diff
@@ -187,6 +187,16 @@ class UserDTO
+    /**
+     * User role
+     * @var string
+     * @Serializer\Type("string")
```
A user can have multiple roles.

SpartakusMd commented on src/DTO/UserDTO.php:
```diff
@@ -197,4 +207,15 @@ public function isLocationValid(ExecutionContext $context): void
+    public function isRoleValid(ExecutionContext $context): void
+    {
+        if (!in_array($this->roles, User::ROLES, true)) {
+            $context->addViolation('Please enter a valid ROLE!', array());
```
The field for which this violation is, was not specified.

SpartakusMd commented on src/Controller/UserController.php:
```diff
@@ -764,8 +796,36 @@ public function setProjectManagerForUser(
+
+        $data = $request->getContent();
+        /** @var DeserializationContext $context */
+        $context = DeserializationContext::create()->setGroups(array('SetRole'));
```
Why is the group called `SetRole`?

SpartakusMd commented on src/Controller/UserController.php:
```diff
@@ -764,8 +796,36 @@ public function setProjectManagerForUser(
+            );
+        }
+
         if ($projectManager->getRoles() === array('ROLE_USER')) {
```
This isn't how you validate if a user has a specific role. Also, you should check specificaly for ROLE_PM as there could be roles like DEV, QA and others.

SpartakusMd commented on src/Controller/UserController.php:
```diff
@@ -764,8 +796,36 @@ public function setProjectManagerForUser(
+            return new JsonResponse(
+                [
+                    'code' => Response::HTTP_BAD_REQUEST,
+                    'message' => 'This user in not a PM!'
```
The message is bad. It says that the user to be changed is not a PM, not that the current user is not a PM.

SpartakusMd commented on src/Controller/UserController.php:
```diff
@@ -764,8 +796,36 @@ public function setProjectManagerForUser(
+            return new JsonResponse(
+                [
+                    'code' => Response::HTTP_NOT_FOUND,
+                    'message' => 'Not found!'
```
What is not found?

Commit:
[BE] Project manager role
https://trello.com/c/mY9KaA6l/99-be-project-manager-role

- Changed endpoint set_pm to set_role
- The ID of PM is in the body not in the URL

Commit:
[BE] Project manager role
https://trello.com/c/mY9KaA6l/99-be-project-manager-role

Commit:
[BE] Project manager role
https://trello.com/c/mY9KaA6l/99-be-project-manager-role

SpartakusMd commented on src/Controller/UserController.php:
```diff
@@ -812,17 +814,17 @@ public function setProjectManagerForUser(
-
-        if ($projectManager->getRoles() === array('ROLE_USER')) {
+        $projectManagerRole = $projectManager->getRoles();
+        if ($projectManagerRole !== array('ROLE_PM') || $projectManagerRole !== array('ROLE_ADMIN')) {
```
You should be using a specific Symfony function to check the role. As ROLE_ADMIN ha already ROLE_PM you won't have to specify it.

Commit:
[BE] Project manager role
https://trello.com/c/mY9KaA6l/99-be-project-manager-role
